File: src/csharp/Foundgine.Core/Semantic/Query/SemanticQueryOptionsValidator.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Query;

/// <summary>
/// Validates query controls before provider planning. Provider-specific
/// limits remain provider policy; semantic invariants are enforced here.
/// </summary>
public static class SemanticQueryOptionsValidator
{
    public static void Validate(SemanticQueryOptions? options, SemanticEntity root)
    {
        ArgumentNullException.ThrowIfNull(root);
        if (options is null) return;

        if (options.Limit is < 0)
            throw new InvalidOperationException("Semantic query limit must be non-negative.");
        if (options.Offset is < 0)
            throw new InvalidOperationException("Semantic query offset must be non-negative.");
        if (options.After is not null && string.IsNullOrWhiteSpace(options.After))
            throw new InvalidOperationException("Semantic cursor cannot be empty.");
        if (options.After is not null && options.Offset is not null)
            throw new InvalidOperationException("Cursor pagination cannot be combined with offset pagination.");
        if (options.After is not null && options.Limit is not > 0)
            throw new InvalidOperationException("Cursor pagination requires a positive limit.");

        ValidateOrderTerms(options.EffectiveOrder, root);
    }

    private static void ValidateOrderTerms(IReadOnlyList<SemanticOrderTerm> terms, SemanticEntity root)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (var term in terms)
        {
            var key =
                $"{string.Join('/', term.EffectivePath.Select(x => x.Value))}:{term.Field.Value}:{term.Direction}:{term.Aggregate}";
            if (!seen.Add(key))
                throw new InvalidOperationException("Semantic query ordering contains a duplicate term.");
        }
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/query/SemanticQueryOptionsValidator.java
package com.foundgine.core.semantic.query;

import com.foundgine.core.semantic.SemanticEntity;
import java.util.*;

/** Validates query controls before provider planning. */
public final class SemanticQueryOptionsValidator {
	private SemanticQueryOptionsValidator() {
	}

	public static void validate(SemanticQueryOptions options, SemanticEntity root) {
		Objects.requireNonNull(root);
		if (options == null)
			return;
		if (options.limit() != null && options.limit() < 0)
			throw invalid("Semantic query limit must be non-negative.");
		if (options.offset() != null && options.offset() < 0)
			throw invalid("Semantic query offset must be non-negative.");
		if (options.after() != null && options.after().isBlank())
			throw invalid("Semantic cursor cannot be empty.");
		if (options.after() != null && options.offset() != null)
			throw invalid("Cursor pagination cannot be combined with offset pagination.");
		if (options.after() != null && (options.limit() == null || options.limit() <= 0))
			throw invalid("Cursor pagination requires a positive limit.");
		var seen = new HashSet<String>();
		for (var term : options.effectiveOrder()) {
			var key = term.path().stream().map(Object::toString).reduce("", (a, b) -> a + (a.isEmpty() ? "" : "/") + b)
					+ ":" + term.field() + ":" + term.direction() + ":" + term.aggregate();
			if (!seen.add(key))
				throw invalid("Semantic query ordering contains a duplicate term.");
		}
	}

	private static IllegalStateException invalid(String m) {
		return new IllegalStateException(m);
	}
}
